[FEATURE] Restreindre l'accès à la page "Équipe" de Pix Admin uniquement au rôle "SUPER_ADMIN" (PIX-4850)

igorissen opened the pull request:
## :unicorn: Problème
La page `Équipe` dans l'application `Pix Admin` est visible à tous les utilisateurs de l'application indépendamment de leur rôles. Seul des utilisateurs ayant le rôle `SUPER_ADMIN` devrait avoir accès à cette page.

![Screenshot 2022-05-06 at 12 39 24](https://user-images.githubusercontent.com/6919604/167116574-7d151b85-beca-41e9-b725-89997efe44fb.png)

## :robot: Solution

1. (API) Mettre à jour la liste des pre-handlers pour n'accepter que les requêtes faites par un utilisateur ayant le rôle `SUPER_ADMIN`.
2. (PIX ADMIN) Ne pas afficher l'onglet `Équipe` pour tous les utilisateurs n'ayant pas le rôle `SUPER_ADMIN`.
3. (PIX-ADMIN) Rediriger l'utilisateur n'ayant pas le rôle `SUPER_ADMIN` à la racine de l'application si ce dernier tente de rentrer le chemin vers la page `/equipe` directement dans l'URL.

## :rainbow: Remarques

Les routes concernées

- API
  - `GET /api/admin/admin-members`
  - `PATCH /api/admin/admin-members/{id}`
- PIX ADMIN
  - `/equipe`

## :100: Pour tester

#### Avec un utilisateur ayant le rôle `SUPER_ADMIN`

- Se connecter avec `superadmin@example.net` / `pix123`
- Voir au niveau du menu à gauche que l'onglet `Équipe` est présent
- Remplacer le chemin `/organizations/list` dans la barre de navigation du navigateur par `/equipe`
- Constater que l'on arrive sur la page `Équipe`
- Modifier le rôle d'un utilisateur **différent** du votre
- Constater dans la liste que le nouveau rôle a été pris en compte

#### Avec un utilisateur ayant le rôle `SUPPORT`

- Se connecter avec `pixsupport@example.net` / `pix123`
- Voir au niveau du menu à gauche que l'onglet `Équipe` n'est pas présent
- Remplacer le chemin `/organizations/list` dans la barre de navigation du navigateur par `/equipe`
- Constater que l'on arrive sur la page `Organisations`

#### Avec un utilisateur ayant le rôle `CERTIF`

- Se connecter avec `pixcertif@example.net` / `pix123`
- Voir au niveau du menu à gauche que l'onglet `Équipe` n'est pas présent
- Remplacer le chemin `/organizations/list` dans la barre de navigation du navigateur par `/equipe`
- Constater que l'on arrive sur la page `Organisations`

#### Avec un utilisateur ayant le rôle `METIER`

- Se connecter avec `pixmetier@example.net` / `pix123`
- Voir au niveau du menu à gauche que l'onglet `Équipe` n'est pas présent
- Remplacer le chemin `/organizations/list` dans la barre de navigation du navigateur par `/equipe`
- Constater que l'on arrive sur la page `Organisations`

igorissen committed:
feat(api): restrict access for GET /api/admin/admin-members

Only an admin member with a role "SUPER_ADMIN" can call this route.

pix-service commented:
I'm deploying this PR to these urls:

- App (.fr): https://app-pr4406.review.pix.fr
- App (.org): https://app-pr4406.review.pix.org
- Orga: https://orga-pr4406.review.pix.fr
- Certif: https://certif-pr4406.review.pix.fr
- Admin: https://admin-pr4406.review.pix.fr
- API: https://api-pr4406.review.pix.fr/api/

 Please check it out!

igorissen committed:
feat(api): restrict access for PATCH /api/admin/admin-members/{id}

Only an admin member with a role "SUPER_ADMIN" can call this route.

igorissen committed:
feat(admin): restrict access to `Équipe` page

Only an admin member with a role "SUPER_ADMIN" can see the tab "Équipe" and open the page.

igorissen committed:
feat(admin): restrict route access to `/equipe`

Only an admin member with a role "SUPER_ADMIN" can access this route.

Anne-Gaelle-S commented on admin/tests/acceptance/authenticated/team/list_test.js:
```diff
@@ -110,4 +112,43 @@ module('Acceptance | Team | List', function (hooks) {
     });
   });
+
+  module('When admin member is logged in with "SUPPORT" as role', function () {
+    test('it should not be accessible', async function (assert) {
+      // given
+      await authenticateAdminMemberWithRole({ isSupport: true })(server);
+
+      // when
+      await visit('/equipe');
+
+      // then
+      assert.strictEqual(currentURL(), '/organizations/list');
+    });
+  });
+
+  module('When admin member is logged in with "CERTIF" as role', function () {
+    test('it should not be accessible', async function (assert) {
+      // given
+      await authenticateAdminMemberWithRole({ isCertif: true })(server);
+
+      // when
+      await visit('/equipe');
+
+      // then
+      assert.strictEqual(currentURL(), '/organizations/list');
+    });
+  });
+
+  module('When admin member is logged in with "METIER" as role', function () {
+    test('it should not be accessible', async function (assert) {
+      // given
+      await authenticateAdminMemberWithRole({ isMetier: true })(server);
+
+      // when
+      await visit('/equipe');
+
+      // then
+      assert.strictEqual(currentURL(), '/organizations/list');
+    });
```
**suggestion**
plutôt que de tester ça en acceptance (parce que les tests front d'acceptance prennent beaucoup de temps) tu peux tester ton ajout de beforeModel dans la route en unitaire : 
[comme ça par exemple](https://github.com/1024pix/pix/pull/4405/commits/bbd594eb35843690f23684627eaa00fee5e36083)

Anne-Gaelle-S commented on admin/tests/integration/components/menu-bar_test.js:
```diff
@@ -14,12 +14,54 @@ module('Integration | Component | menu-bar', function (hooks) {
+      assert.dom(screen.getByTitle('Équipe')).exists();
+    });
+
+    test('should not contain link to "team" management page when admin member have "SUPPORT" as role', async function (assert) {
+      // given
+      const currentUser = this.owner.lookup('service:currentUser');
+      currentUser.adminMember = { isSupport: true };
+
+      // when
+      const screen = await render(hbs`{{menu-bar}}`);
+
+      // then
+      assert.throws(() => screen.getByTitle('Équipe'), 'HTMLElement not found');
+    });
+
+    test('should not contain link to "team" management page when admin member have "CERTIF" as role', async function (assert) {
+      // given
+      const currentUser = this.owner.lookup('service:currentUser');
+      currentUser.adminMember = { isCertif: true };
+
+      // when
+      const screen = await render(hbs`{{menu-bar}}`);
+
+      // then
+      assert.throws(() => screen.getByTitle('Équipe'), 'HTMLElement not found');
+    });
+
+    test('should not contain link to "team" management page when admin member have "METIER" as role', async function (assert) {
+      // given
+      const currentUser = this.owner.lookup('service:currentUser');
+      currentUser.adminMember = { isMetier: true };
+
+      // when
+      const screen = await render(hbs`{{menu-bar}}`);
+
+      // then
+      assert.throws(() => screen.getByTitle('Équipe'), 'HTMLElement not found');
+    });
```
**suggestion**
Tu peux fusionner ces 3 tests en un : 
```suggestion
    test('should not contain link to "team" management page when admin member have support, metier or certif role', async function (assert) {
      // given
      const currentUser = this.owner.lookup('service:currentUser');
      currentUser.adminMember = { isSuperAdmin: false };

      // when
      const screen = await render(hbs`{{menu-bar}}`);

      // then
      assert.throws(() => screen.getByTitle('Équipe'), 'HTMLElement not found');
    });
 ```

Anne-Gaelle-S commented on api/tests/acceptance/application/admin-members/admin-members-route-get_test.js:
```diff
@@ -29,10 +29,73 @@ describe('Acceptance | Application | Admin-members | Routes', function () {
+      // given
+      const user = databaseBuilder.factory.buildUser();
+      databaseBuilder.factory.buildPixAdminRole({ userId: user.id, role: ROLES.CERTIF });
+      const admin = await insertAdminMemberWithRole({ role: ROLES.CERTIF });
+      await databaseBuilder.commit();
+      const server = await createServer();
+
+      // when
+      const response = await server.inject({
+        headers: {
+          authorization: generateValidRequestAuthorizationHeader(admin.id),
+        },
+        method: 'GET',
+        url: `/api/admin/admin-members`,
+      });
+
+      // then
+      expect(response.statusCode).to.equal(403);
+    });
+
+    it('should return 403 http status code when user has role "METIER"', async function () {
+      // given
+      const user = databaseBuilder.factory.buildUser();
+      databaseBuilder.factory.buildPixAdminRole({ userId: user.id, role: ROLES.METIER });
+      const admin = await insertAdminMemberWithRole({ role: ROLES.METIER });
+      await databaseBuilder.commit();
+      const server = await createServer();
+
+      // when
+      const response = await server.inject({
+        headers: {
+          authorization: generateValidRequestAuthorizationHeader(admin.id),
+        },
+        method: 'GET',
+        url: `/api/admin/admin-members`,
+      });
+
+      // then
+      expect(response.statusCode).to.equal(403);
+    });
```
**suggestion**
Idem ici, pour limiter les tests d'acceptance et faire de l'unitaire first , je propose de tester tes modifications sur la route en unitaire : [exemple ici](https://github.com/1024pix/pix/pull/4405/commits/2605e8a0e94ed77b6066af37ba89efde85473b97)

Anne-Gaelle-S commented on api/tests/acceptance/application/admin-members/admin-members-route-get_test.js:
```diff
@@ -63,6 +126,99 @@ describe('Acceptance | Application | Admin-members | Routes', function () {
       expect(response.statusCode).to.equal(200);
     });
 
+    it('should return 403 http status code if use has role "SUPPORT"', async function () {
```
**suggestion**
Idem pour ici, les tests d'acceptance sont couteux alors si on peut tester les modifs de pré-handler autre part et laisser les tests d'acceptance vérifier le "happy path" ou vérifier que la tuyauterie passe bien je trouve ça mieux

Anne-Gaelle-S commented on api/tests/test-helper.js:
```diff
@@ -59,6 +59,21 @@ function generateIdTokenForExternalUser(externalUser) {
   return tokenService.createIdTokenForUserReconciliation(externalUser);
 }
 
+async function insertAdminMemberWithRole({ role } = {}) {
+  const user = databaseBuilder.factory.buildUser.withRole({
+    id: 1234,
+    firstName: 'Super',
+    lastName: 'Papa',
+    email: 'super.papa@example.net',
+    password: 'Password123',
+    role,
+  });
+
+  await databaseBuilder.commit();
+
+  return user;
```
**question**
Pourquoi avoir rajouté cette fonction dans les helpers plutôt que mettre directement l'appel à la ligne 16 dans `api/tests/acceptance/application/admin-members/admin-members-route-get_test.js` ?

igorissen replied:
C'était pour éviter d'écrire plusieurs fois les 2 actions (créer un utilisateur avec un role et valider l'insertion en db), et aussi pour éviter de modifier la méthode `insertUserWithRoleSuperAdmin` qui beaucoup utilisé (petit commit).

clemlatz replied:
Je suis un peu sceptique aussi sur cette factorisation car cette méthode `insertAdminMember` ne fait qu'appeler une autre méthode du databasebuilder en 
- ajoutant des propriétés à l'user (mais il me semble qu'on pourrait s'en passer, on ne s'en sert pas dans le test)
- faisait le commit builder (mais à mon avis c'est mieux que ça soit fait dans le test, l'idée étant de n'en faire qu'un s'il y a plusieurs insertions)

Anne-Gaelle-S commented on api/tests/acceptance/application/admin-members/admin-members-route-get_test.js:
```diff
@@ -2,18 +2,18 @@ const {
       // given
       const user = databaseBuilder.factory.buildUser();
-      databaseBuilder.factory.buildPixAdminRole({ userId: user.id, role: 'SUPER_ADMIN' });
-      const admin = await insertUserWithRoleSuperAdmin();
+      databaseBuilder.factory.buildPixAdminRole({ userId: user.id, role: ROLES.SUPER_ADMIN });
+      const admin = await insertAdminMemberWithRole({ role: ROLES.SUPER_ADMIN });
```
**question**
J'ai l'impression que les lignes 14 et 15 ne servent plus (`user` n'est utile nulle part dans ce test ?). Est-ce qu'elles ont encore une utilités ?

Anne-Gaelle-S commented on admin/tests/unit/routes/authenticated/team/list_test.js:
```diff
@@ -0,0 +1,48 @@
+  test('it should allow super admin to access page', function (assert) {
+    // given
+    const route = this.owner.lookup('route:authenticated/team/list');
+    assert.ok(route);
```
L'assert ici a été oublié je pense : 
```suggestion
```

Anne-Gaelle-S commented on admin/tests/unit/routes/authenticated/team/list_test.js:
```diff
@@ -0,0 +1,48 @@
+  test('it should redirect all other pix member (certif, metier, support) to home page', function (assert) {
+    // given
+    const route = this.owner.lookup('route:authenticated/team/list');
+    assert.ok(route);
```
```suggestion
```
